### PR TITLE
Fix/s3 error fallback

### PR DESCRIPTION
**Pull request by ianopolous**

### CODE REVIEW

File: src/peergos/shared/storage/RetryStorage.java [hunk 1]
```diff
@@ -42,28 +41,30 @@ private int jitter(int minMilliseconds, int rangeMilliseconds) {
...
-                i-> {
-                    CompletableFuture<V> res = new CompletableFuture<>();
-                    f.get()
-                            .thenAccept(res::complete)
-                            .exceptionally(e ->  {
-                                if (i == 1) {
-                                    res.completeExceptionally(e);
-                                } else {
-                                    retryAfter(() -> recurse(i - 1, f)
+        CompletableFuture<V> res = new CompletableFuture<>();
```
Comment (kevodwyer) [inline src/peergos/shared/storage/RetryStorage.java:44]: how about moving CF res up a level into runWithRetry and passing it through as a method param?

File: src/peergos/shared/storage/RetryStorage.java [hunk 2]
```diff
@@ -42,28 +41,30 @@ private int jitter(int minMilliseconds, int rangeMilliseconds) {
...
+        try {
+            f.get()
+                    .thenAccept(res::complete)
+                    .exceptionally(e -> {
+                        if (retriesLeft == 1) {
+                            res.completeExceptionally(e);
+                        } else if (e instanceof HttpFileNotFoundException) {
```
Comment (kevodwyer) [inline src/peergos/shared/storage/RetryStorage.java:51]: That's smart!

File: src/peergos/shared/storage/DirectS3BlockStore.java [hunk 3]
```diff
@@ -211,7 +212,17 @@ public static Multihash keyToHash(String keyFileName) {
                             res[i] = new FragmentWithHash(new Fragment(identity.getHash()), Optional.empty());
                         }
                     return Arrays.asList(res);
+                }).thenAccept(allResults::complete)
+                .exceptionally(t -> {
+                    NetworkAccess.downloadFragments(hashes, this, monitor, spaceIncreaseFactor)
```
Comment (kevodwyer) [inline src/peergos/shared/storage/DirectS3BlockStore.java:217]: Good to add a comment here to explain fallback

Comment (ianopolous) [reply src/peergos/shared/storage/DirectS3BlockStore.java:217]: Is it not clear given it is the same as the call at the beginning of the method?

File: src/peergos/shared/storage/RetryStorage.java (hunk 1 above)
Comment (ianopolous) [reply src/peergos/shared/storage/RetryStorage.java:44]: I think it's cleaner to not have references being passed around.

File: src/peergos/shared/storage/DirectS3BlockStore.java (hunk 3 above)
Comment (kevodwyer) [reply src/peergos/shared/storage/DirectS3BlockStore.java:217]: not that bothered

File: src/peergos/shared/storage/RetryStorage.java (hunk 1 above)
Comment (kevodwyer) [reply src/peergos/shared/storage/RetryStorage.java:44]: ok

**Pull request merged**